pkgs/development/php-packages/castor/default.nix

```diff
@@ -10,28 +9,23 @@
 
 php.buildComposerProject (finalAttrs: {
   pname = "castor";
-  version = "0.14.0";
+  version = "0.17.1";
 
   src = fetchFromGitHub {
     owner = "jolicode";
     repo = "castor";
     rev = "v${finalAttrs.version}";
-    hash = "sha256-sSIkXNW6RR1mx15dKouQLMaHBr5FEkTTc/0QIkWV8sg=";
+    hash = "sha256-ng32vuGlGffpkzf3hXu0sNbj0PCDu4DpZnMnbDV9pZk=";
   };
 
-  patches = [
-    # Upstream lock is invalid. https://github.com/jolicode/castor/issues/319
-    (fetchpatch {
-      name = "fix-invalid-lock.patch";
-      url = "https://github.com/jolicode/castor/commit/5ff0c3ecbdddad20146adbc2f055b83f5aadba0f.patch";
-      hash = "sha256-1a3Dpk/UXp92Ugw9zSoLPsbWOJEuP2FBWc/pQ/EKwaM=";
-    })
-  ];
-
-  vendorHash = "sha256-HfEjwlkozeuT4LDnYwiCu7T0spcf4GLhkd7Kc1VRnro=";
+  vendorHash = "sha256-E2NMWuUQXQ96NbKrcFnFGlxWR0tkd56MXk1bCL0N/sE=";
 
   nativeBuildInputs = [ installShellFiles ];
 
+  postPatch = ''
+    patchShebangs bin/castor
```

Comment:
Why ?!

Reply:
Because otherwise I would get:
```
       > /nix/store/6a7qxqcm33y3lwr1v55hpp5hg7akja92-stdenv-linux/setup: line 176: /nix/store/01vv85bd1f64hshhvyj99bfjc9b5iz8j-castor-0.17.1/bin/castor: cannot execute: required file not found
       > /nix/store/6a7qxqcm33y3lwr1v55hpp5hg7akja92-stdenv-linux/setup: line 176: /nix/store/01vv85bd1f64hshhvyj99bfjc9b5iz8j-castor-0.17.1/bin/castor: cannot execute: required file not found
       > /nix/store/6a7qxqcm33y3lwr1v55hpp5hg7akja92-stdenv-linux/setup: line 176: /nix/store/01vv85bd1f64hshhvyj99bfjc9b5iz8j-castor-0.17.1/bin/castor: cannot execute: required file not found
       > installShellCompletion: error: installed shell completion file `/nix/store/01vv85bd1f64hshhvyj99bfjc9b5iz8j-castor-0.17.1/share/bash-completion/completions/castor.bash' does not exist or has zero size
```

Similar fix was done for phpactor in https://github.com/NixOS/nixpkgs/pull/326091

Reply:
But isn't this supposed to be automatically done in the `fixup` phase?

Reply:
Here is the hydra log: https://hydra.nixos.org/build/264938630/nixlog/4 for one of the previous runs. It doesn't seem fixup is run.

Reply:
I'm comparing phpactor hydra https://hydra.nixos.org/build/268320955/log and fixup happens after postInstall. Maybe this is the problem.

Reply:
Normally this should happen in postinstall yeah

Reply:
I dug out the failing build for phpactor - before the postPatch was added and it failed in the same way as castor: https://hydra.nixos.org/build/264952873/nixlog/4

Reply:
We could remove it for both packages when we have a proper fix. I'm not sure if this should prevent unbreaking the build.

Reply:
OK.

Reply:
Could you add a comment, indicating that we should investigate and find why the postInstall step is not triggered?

Reply:
It seems to be related to https://github.com/NixOS/nixpkgs/pull/289517

Reply:
@SuperSandro2000 Do you have a clue?

Reply:
@drupol I added the comment for now.

Reply:
@jtojnar Do you have a clue for this? (since you're into PHP!)

Reply:
Nothing PHP-specific. `<($out/bin/castor completion bash)` runs the command, [redirects the output](https://www.gnu.org/software/bash/manual/html_node/Process-Substitution.html) to a file descriptor, and returns path to the file descriptor.

Since the command is run in `postInstall` hook, `patchShebangsAuto` has not been executed yet (it runs in `fixupPhase`), so `$out/bin/castor` still has a shebang that does not exist in the sandbox and will fail to execute.

When the command fails to execute, the output will be empty. https://github.com/NixOS/nixpkgs/pull/289517 now throws an error in such case to catch the completions being broken.

---

Running `patchShebangs` manually earlier is a reasonable solution. Though you should pass it `--host` flag so that continues to use correct (host) `PATH` when `strictDeps` are enabled.

I would also move it to `postInstall` since the patching is not really needed before then.

Reply:
@jtojnar what do you think about moving it to `fixupPhase` like in https://github.com/NixOS/nixpkgs/pull/331782/files

Reply:
Actually I tried it locally and it doesn't seem to work.

Reply:
That PR is broken. `fixupPhase` does [quite a lot of stuff](https://github.com/jtojnar/nixpkgs-hammering/blob/main/explanations/fixup-phase.md) and when they override it like that, it will no longer happen. If it really needs to go to `fixupPhase`, it should be put it in `postFixup` hook.

But IMO, installing setup hooks makes more sense in `installPhase`.

Alternately, if you want to avoid manual patching of shebangs, calling the interpreter explicitly will probably be even cleaner  `<(php $out/bin/castor completion bash)`